File: src/main/java/me/yamakaja/commanditems/data/action/ActionMessage.java
package me.yamakaja.commanditems.data.action;

import java.util.List;
import java.util.logging.Level;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;

import com.fasterxml.jackson.annotation.JsonProperty;

import me.yamakaja.commanditems.CommandItems;
import me.yamakaja.commanditems.data.ItemDefinition;
import me.yamakaja.commanditems.interpreter.InterpretationContext;

public class ActionMessage extends Action {

    private MessageTarget target;
    private String message;
    private String permission;

    public ActionMessage(@JsonProperty("action") ActionType type, @JsonProperty("to") MessageTarget target, @JsonProperty(value = "message", required = true) String message, @JsonProperty("perm") String permission) {
        super(type);
        this.target = target;

        if (target == null)
            this.target = MessageTarget.PLAYER;

        this.message = ChatColor.translateAlternateColorCodes('&', message);
        this.permission = permission;
    }

    @Override
    public void trace(List<ItemDefinition.ExecutionTrace> trace, int depth) {
        String line;

        switch (this.target) {
            case PLAYER:
                line = String.format("To player: %s", this.message);
                break;
            case CONSOLE:
                line = String.format("To console: %s", this.message);
                break;
            case EVERYBODY:
                line = String.format("To everybody: %s", this.message);
                break;
            case PERMISSION:
                line = String.format("To everybody with permission %s: %s", this.permission, this.message);
                break;
            default:
                throw new IllegalStateException("Unexpected trace value: " + this.target);
        }

        trace.add(new ItemDefinition.ExecutionTrace(depth, line));
    }

    @Override
    public void process(InterpretationContext context) {
        String message = context.resolveLocalsInString(this.message);
        switch (this.target) {
            case PLAYER:
                context.getPlayer().sendMessage(message);
                break;

            case CONSOLE:
                context.getPlugin().getLogger().info("[MSG] " + message);
                break;

            case EVERYBODY:
                Bukkit.getServer().broadcastMessage(message);
                break;

            case PERMISSION:
                if (permission == null)
                    CommandItems.logger.log(Level.SEVERE, "[CMDI] Permission is null in permission node!");

                Bukkit.getOnlinePlayers().stream()
                        .filter(player -> player.hasPermission(this.permission))
                        .forEach(player -> player.sendMessage(message));
                break;
            default:
                throw new IllegalStateException("Unexpected process value: " + this.target);
        }
    }

    /**
     * Created by Yamakaja on 26.05.18.
     */
    public enum MessageTarget {

        PLAYER,
        CONSOLE,
        EVERYBODY,
        PERMISSION

    }
}

File: src/main/kotlin/me/relaxing9/commanditems/data/action/ActionMessage.kt
package me.relaxing9.commanditems.data.action

import com.fasterxml.jackson.annotation.JsonProperty
import me.relaxing9.commanditems.CommandItems
import me.relaxing9.commanditems.data.ItemDefinition.ExecutionTrace
import me.relaxing9.commanditems.interpreter.InterpretationContext
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import java.util.logging.Level

class ActionMessage(
    @JsonProperty("action") type: ActionType?,
    @param:JsonProperty("to") private val target: MessageTarget?,
    @JsonProperty(value = "message", required = true) message: String?,
    @JsonProperty("perm") permission: String?
) :
    Action(type!!) {
    private val message: String
    private val permission: String?

    init {
        if (target == null) target = MessageTarget.PLAYER
        this.message = ChatColor.translateAlternateColorCodes('&', message!!)
        this.permission = permission
    }

    override fun trace(trace: MutableList<ExecutionTrace>, depth: Int) {
        val line: String
        when (target) {
            MessageTarget.PLAYER -> line = String.format("To player: %s", message)
            MessageTarget.CONSOLE -> line = String.format("To console: %s", message)
            MessageTarget.EVERYBODY -> line = String.format("To everybody: %s", message)
            MessageTarget.PERMISSION -> line = String.format(
                "To everybody with permission %s: %s",
                permission, message
            )

            else -> throw IllegalStateException("Unexpected trace value: $target")
        }
        trace.add(ExecutionTrace(depth, line))
    }

    override fun process(context: InterpretationContext) {
        val message = context.resolveLocalsInString(message)
        when (target) {
            MessageTarget.PLAYER -> context.player.sendMessage(message)
            MessageTarget.CONSOLE -> context.plugin.logger.info("[MSG] $message")
            MessageTarget.EVERYBODY -> Bukkit.getServer().broadcastMessage(message)
            MessageTarget.PERMISSION -> {
                if (permission == null) CommandItems.logger.log(
                    Level.SEVERE,
                    "[CMDI] Permission is null in permission node!"
                )
                Bukkit.getOnlinePlayers().stream()
                    .filter { player: Player? -> player!!.hasPermission(permission!!) }
                    .forEach { player: Player? -> player!!.sendMessage(message) }
            }

            else -> throw IllegalStateException("Unexpected process value: $target")
        }
    }

    /**
     * Created by Yamakaja on 26.05.18.
     */
    enum class MessageTarget {
        PLAYER, CONSOLE, EVERYBODY, PERMISSION
    }
}